### DISCUSSION TITLE
Cancel custom action from bundle

### DISCUSSION THREAD

**Discussion by gbull122** [wixtoolset/issues#8006] [Questions] [answered]
I have a long running custom action in C# and it is done in several steps so I use `session.Message(InstallMessage.Progress, record); ` to report progress. Running the MSI and clicking cancel stops the installer and I can see from the logs that the custom action is stopped on the relevant step, all good.

I have to put the MSI into a bundle that doesn't  show the MSI UI. Now clicking cancel on the bundle UI stops the install but allows the custom action to complete all steps. Is is possible to get the bundle to cancel the custom action?

**Comment by robmen** [comment]
Yes. If your BA responds with cancel on the progress and MSI callbacks, that should work.

**Reply by gbull122** [reply]
Ok, that means I need my own BA then doesn't it? 
I was hoping I was missing something that was already in the standard Wix BA.

**Reply by robmen** [reply]
I would expect wixstdba to do cancel.

**Reply by gbull122** [reply, marked as answer]
Thanks @robmen, you've helped me answer my own question.

Knowing it should work made me investigate the [ProcessMessage](https://learn.microsoft.com/en-us/windows/win32/api/msiquery/nf-msiquery-msiprocessmessage) documentation.  Now in my C# custom action I set the record of the message to make it "Explicit"

`var recordProgress = new Record(3);
recordProgress.SetInteger(1, 1);
recordProgress.SetInteger(2, 1);
recordProgress.SetInteger(3, 0);

session.Message(InstallMessage.Progress, recordProgress);`

and clicking the cancel button in the bundle UI does indeed cancel the action.
